impl_core/src/main/java/io/opencensus/implcore/metrics/MetricRegistryImpl.java

```diff
@@ -46,56 +43,38 @@
   }
 
   @Override
-  public <T> void addLongGauge(
-      String name,
-      String description,
-      String unit,
-      LinkedHashMap<LabelKey, LabelValue> labels,
-      T obj,
-      ToLongFunction<T> function) {
-    checkNotNull(labels, "labels");
-    registeredMeters.registerMeter(
-        new LongGauge<T>(
+  public LongGauge addLongGauge(
+      String name, String description, String unit, List<LabelKey> labelKeys) {
+    Utils.checkListElementNotNull(
+        checkNotNull(labelKeys, "labelKeys should not be null."),
+        "labelKeys element should not be null.");
+    LongGaugeImpl longGaugeMetric =
+        new LongGaugeImpl(
             checkNotNull(name, "name"),
             checkNotNull(description, "description"),
             checkNotNull(unit, "unit"),
-            Collections.unmodifiableList(new ArrayList<LabelKey>(labels.keySet())),
-            Collections.unmodifiableList(new ArrayList<LabelValue>(labels.values())),
-            obj,
-            checkNotNull(function, "function")));
-  }
-
-  @Override
-  public <T> void addDoubleGauge(
-      String name,
-      String description,
-      String unit,
-      LinkedHashMap<LabelKey, LabelValue> labels,
-      T obj,
-      ToDoubleFunction<T> function) {
-    checkNotNull(labels, "labels");
-    registeredMeters.registerMeter(
-        new DoubleGauge<T>(
-            checkNotNull(name, "name"),
-            checkNotNull(description, "description"),
-            checkNotNull(unit, "unit"),
-            Collections.unmodifiableList(new ArrayList<LabelKey>(labels.keySet())),
-            Collections.unmodifiableList(new ArrayList<LabelValue>(labels.values())),
-            obj,
-            checkNotNull(function, "function")));
+            Collections.unmodifiableList(new ArrayList<LabelKey>(labelKeys)));
+    registeredMeters.registerMeter(name, longGaugeMetric);
+    return longGaugeMetric;
   }
 
   private static final class RegisteredMeters {
-    private volatile Set<Gauge> registeredGauges = Collections.emptySet();
+    private volatile Map<String, Meter> registeredMeters = Collections.emptyMap();
 
-    private Set<Gauge> getRegisteredMeters() {
-      return registeredGauges;
+    private Map<String, Meter> getRegisteredMeters() {
+      return registeredMeters;
     }
 
-    private synchronized void registerMeter(Gauge gauge) {
-      Set<Gauge> newGaguesList = new LinkedHashSet<Gauge>(registeredGauges);
-      newGaguesList.add(gauge);
-      registeredGauges = Collections.unmodifiableSet(newGaguesList);
+    private synchronized void registerMeter(String meterName, Meter meter) {
+      Meter existingMeter = registeredMeters.get(meterName);
+      if (existingMeter != null) {
```

Comment:
If `existingMeter` and `meter` is same, this should be a no-op instead of throwing exception.

Reply:
I think ```Meter``` interface should expose ```MetricDescriptor``` to do this check OR another way is to override ```equals``` and ```hashCode``` methods in all Meter classes (like LongGauge, DoubleGauge etc.)

Reply:
Could you add a TODO on this? Not a critical issue but I think we should allow users to register the same `Meter` multiple times. (Otherwise we should document the behavior that registering the same `Meter` for more than once will throw an exception.)